src/oidc_child/oidc_child.c

```diff
@@ -0,0 +1,1396 @@
+        poptPrintUsage(pc, stderr, 0);
+        poptFreeContext(pc);
+        _exit(-1);
+    }
+
+    if (get_device_code && get_access_token) {
+        fprintf(stderr,
+                "\n--get-device-code and --get-access-token "
+                "are mutually exclusive .\n\n");
+        poptPrintUsage(pc, stderr, 0);
+        poptFreeContext(pc);
+        _exit(-1);
+    }
+
+    if ((issuer_url != NULL && (issuer != NULL
+                                || device_auth_endpoint != NULL
+                                || token_endpoint != NULL))
+        || (issuer != NULL && (issuer_url != NULL
+                                || device_auth_endpoint != NULL
+                                || token_endpoint != NULL))
+        || (device_auth_endpoint != NULL && token_endpoint != NULL
+                           && (issuer_url != NULL || issuer != NULL))
+        || (issuer_url == NULL && issuer == NULL
+                && ((device_auth_endpoint != NULL && token_endpoint == NULL)
+                   || (device_auth_endpoint == NULL && token_endpoint != NULL)))
+        || (issuer_url == NULL && issuer == NULL
+                && (device_auth_endpoint == NULL || token_endpoint == NULL))) {
+        fprintf(stderr, "\n--issuer-url or --issuer or --device-auth-endpoint "
+                        "together with --token-endpoint are mutually exclusive "
+                        "but one variant must be given.\n\n");
+        poptPrintUsage(pc, stderr, 0);
+        poptFreeContext(pc);
+        _exit(-1);
+    }
+
+    if (client_id == NULL) {
+        fprintf(stderr, "\n--client must be given.\n\n");
+        poptPrintUsage(pc, stderr, 0);
+        poptFreeContext(pc);
+        _exit(-1);
+    }
+
+    poptFreeContext(pc);
+
+    debug_prg_name = talloc_asprintf(NULL, "oidc_child[%d]", getpid());
+    if (debug_prg_name == NULL) {
+        ERROR("talloc_asprintf failed.\n");
+        goto fail;
+    }
+
+    if (debug_fd != -1) {
+        opt_logger = sss_logger_str[FILES_LOGGER];
+        ret = set_debug_file_from_fd(debug_fd);
+        if (ret != EOK) {
+            opt_logger = sss_logger_str[STDERR_LOGGER];
+            ERROR("set_debug_file_from_fd failed.\n");
+        }
+    }
+
+    DEBUG_INIT(debug_level, opt_logger);
+
+    DEBUG(SSSDBG_TRACE_FUNC, "oidc_child started.\n");
+
+    DEBUG(SSSDBG_TRACE_INTERNAL,
+          "Running with effective IDs: [%"SPRIuid"][%"SPRIgid"].\n",
+          geteuid(), getegid());
+
+    DEBUG(SSSDBG_TRACE_INTERNAL,
+          "Running with real IDs [%"SPRIuid"][%"SPRIgid"].\n",
+          getuid(), getgid());
+
+    main_ctx = talloc_new(NULL);
+    if (main_ctx == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "talloc_new failed.\n");
+        talloc_free(discard_const(debug_prg_name));
+        goto fail;
+    }
+    talloc_steal(main_ctx, debug_prg_name);
+
+    dc_ctx = get_dc_ctx(main_ctx, libcurl_debug, ca_db, issuer_url, issuer,
```

Comment:
I moved it into a separate function, do you think it is worth to create a new file for it?

Reply:
I personally like to split code into separate modules where possible, besides the fact that it can compile fast, it also helps me to read the code. So I would definitely welcomed it, but its my personal opinion so it is up to you.

Reply:
Hi,

I've split out curl and json related code into separate file. Since the file is now much smaller I kept the config parsing function in the main file.

bye,
Sumit